klippy/extras/palette2.py

```diff
@@ -0,0 +1,610 @@
+                self.smart_load_timer = None
+                self.is_loading = False
+
+        def feedrateStart(params):
+            logging.info("Setting feedrate to %f for splice" %
+                         self.feedrate_splice)
+            self.is_splicing = True
+            self.gcode.run_script("M220 S%d" % (self.feedrate_splice * 100))
+
+        def feedrateEnd(params):
+            logging.info("Setting feedrate to %f splice done" %
+                         self.feedrate_normal)
+            self.is_splicing = False
+            self.gcode.run_script("M220 S%d" % (self.feedrate_normal * 100))
+
+        matchers = []
+        if self.is_printing:
+            matchers = matchers + [
+                [printCancelling, 2, "U0", "D2"],
+                [printCancelled, 2, "U0", "D3"],
+                [loadingOffset, 2, "U39"],
+                [loadingOffsetStart, 1, "U39"],
+            ]
+
+        matchers.append([feedrateStart, 3, "U25", "D0"])
+        matchers.append([feedrateEnd, 3, "U25", "D1"])
+        self._param_Matcher(matchers, params)
+
+    def cmd_P2_O100(self, params):
+        logging.info("Pause request from Palette 2")
+        self.is_setup_complete = True
+        self.pause_resume.send_resume_command()
+
+    def cmd_P2_O102(self, params):
+        toolhead = self.printer.lookup_object("toolhead")
+        if not toolhead.get_extruder().get_heater().can_extrude:
+            self.write_queue.put(COMMAND_SMART_LOAD_STOP)
+            self.gcode.respond_info(
+                "Unable to auto load filament, extruder is below minimum temp")
+            return
+
+        if self.smart_load_timer is None:
+            logging.info("Smart load starting")
+            reactor = self.printer.get_reactor()
+            self.smart_load_timer = reactor.register_timer(
+                self._run_Smart_Load, reactor.NOW)
+
+    def cmd_P2(self, line):
+        t = line.split()
+        ocode = t[0]
+        params = t[1:]
+        params_count = len(params)
+        if params_count:
+            res = [i for i in params if i[0] == "D" or i[0] == "U"]
+            if not all(res):
+                logging.error("Omega parameters are invalid")
+                return
+
+        func = getattr(self, 'cmd_P2_' + ocode, None)
+        if func is not None:
+            func(params)
+
+    def _param_Matcher(self, matchers, params):
+        # Match the command with the handling table
+        for matcher in matchers:
+            if len(params) >= matcher[1]:
+                match_params = matcher[2:]
+                res = all([match_params[i] == params[i]
+                           for i in range(len(match_params))])
+                if res:
+                    matcher[0](params)
+                    return True
+
+        return False
+
+    def _run_Read(self, eventtime):
+        while self.serial.in_waiting > 0:
+            try:
+                raw_line = self.serial.readline()
```

Comment:
Unfortunately, this is problematic, as the `readline()` call could block indefinitely - which would result in a hard to debug failure in Klipper.  Generally speaking, the Klipper host code (klippy.py) is intended to only read/write to the mirco-controller code so as to avoid issues like this.

Reply:
Serial is initialized with a timeout, readline is guaranteed to return within about 500ms and it will insert EOL at the end. The device only returns a few bytes at a time so this is more than sufficient headroom, actually could be safely brought down to 250ms.

I could move to threads and use consumer queues on the timers instead, but didn't see other examples of this in the code so was unsure if it was simply too complex for the use case.

Reply:
FWIW, there's a lot of overhead to threads.  Simpler would probably be to set the serial fd to non-blocking mode.

-Kevin

Reply:
Moved to single-threaded fully non-blocking.